File: realtime/ios/webrtc/Sources/Realtime/RealtimeSession.swift
import AVFoundation
import Foundation
import WebRTC

enum SessionState: Equatable {
    case idle
    case connecting
    case connected
    case failed(String)
}

struct SessionConfig {
    enum TurnDetection {
        case semanticVAD(eagerness: String)
        case serverVAD(threshold: Double?, prefixPaddingMs: Int?, silenceDurationMs: Int?, idleTimeoutMs: Int?)
    }

    struct BackchannelConfig {
        var maxPerTurn: Int
        var minGapMs: Int
        var minSpeechMs: Int
        var volumeGain: Double
        var deciderKind: String
        var ruleFireProbability: Double
    }

    struct ResponsivenessConfig {
        var initialWaitTimeoutMs: Int
        var maxInitialPerTurn: Int
        var minFillerGapMs: Int
        var maxTokens: Int
        var enableOnFirstReply: Bool
    }

    var model: String
    var instructions: String
    var temperature: Double?
    var maxOutputTokens: Int?

    var ttsModel: String
    var voice: String
    var speechSpeed: Double?

    var transcriptionModel: String?
    var transcriptionLanguage: String?
    var noiseReduction: String?
    var turnDetection: TurnDetection = .semanticVAD(eagerness: "high")
    var createResponse = true
    var interruptResponse = true

    var backchannel: BackchannelConfig?
    var responsiveness: ResponsivenessConfig?

    var greetingPrompt: String
}

@MainActor
final class RealtimeSession {
    private(set) var state: SessionState = .idle {
        didSet { onStateChange?(state) }
    }

    var onStateChange: ((SessionState) -> Void)?
    var onEvent: ((ServerEvent) -> Void)?

    private let authProvider: AuthProvider
    private let config: SessionConfig
    private var client: WebRTCClient?
    private let backchannelPlayer = BackchannelAudioPlayer()
    private var interrupted = false

    init(authProvider: AuthProvider, config: SessionConfig) {
        self.authProvider = authProvider
        self.config = config
    }

    func connect() async {
        guard state == .idle || isFailed else { return }
        state = .connecting
        do {
            guard await AVAudioApplication.requestRecordPermission() else {
                state = .failed("Microphone access denied. Enable it in iOS Settings.")
                return
            }

            let credentials = try await authProvider.credentials()
            let api = SignalingAPI(credentials: credentials)
            let iceServers = try await api.fetchIceServers()

            let client = try WebRTCClient(iceServers: iceServers)
            self.client = client
            wireCallbacks(client)

            let offerSDP = try await client.makeOfferSDP()
            let answerSDP = try await api.postOffer(sdp: offerSDP)
            try await client.setAnswer(sdp: answerSDP)
            AudioSessionController.overrideToSpeaker()
        } catch {
            disconnect(failure: error.localizedDescription)
        }
    }

    func disconnect(failure: String? = nil) {
        client?.close()
        client = nil
        backchannelPlayer.stop()
        interrupted = false
        AudioSessionController.deactivate()
        state = failure.map(SessionState.failed) ?? .idle
    }

    func setMicEnabled(_ enabled: Bool) {
        client?.setMicEnabled(enabled)
    }

    private var isFailed: Bool {
        if case .failed = state { return true }
        return false
    }

    private func wireCallbacks(_ client: WebRTCClient) {
        client.onDataChannelOpen = { [weak self] in
            Task { @MainActor in self?.sendInitialEvents() }
        }
        client.onServerEvent = { [weak self] event in
            Task { @MainActor in self?.handle(event) }
        }
        client.onConnectionStateChange = { [weak self] pcState in
            Task { @MainActor in self?.handleConnectionState(pcState) }
        }
    }

    private func sendInitialEvents() {
        guard let client else { return }
        client.send(SessionUpdateEvent(config: config))
        client.send(ConversationItemCreateEvent(userText: config.greetingPrompt))
        client.send(ResponseCreateEvent())
        state = .connected
    }

    private func handle(_ event: ServerEvent) {
        switch event {
        case .speechStarted:
            // Barge-in: silence the agent immediately, cancel its response.
            interrupted = true
            client?.setAgentAudioEnabled(false)
            client?.send(ResponseCancelEvent())
        case .outputItemAdded:
            if interrupted {
                client?.setAgentAudioEnabled(true)
                interrupted = false
            }
        case .backchannelAudioDelta(let base64PCM16):
            // Played independently of the WebRTC track so it stays audible during barge-in.
            backchannelPlayer.enqueue(base64PCM16: base64PCM16)
        default:
            break
        }
        onEvent?(event)
    }

    private func handleConnectionState(_ pcState: RTCPeerConnectionState) {
        switch pcState {
        case .failed:
            disconnect(failure: "Connection lost.")
        case .disconnected, .closed:
            if state == .connected { disconnect() }
        default:
            break
        }
    }
}

File: realtime/flutter/webrtc/lib/realtime/realtime_session.dart
import 'package:flutter_webrtc/flutter_webrtc.dart';
import 'package:permission_handler/permission_handler.dart';

import '../audio/backchannel_player.dart';
import '../auth/auth_provider.dart';
import '../webrtc/webrtc_client.dart';
import 'events/client_events.dart';
import 'events/server_event.dart';
import 'session_config.dart';
import 'signaling_api.dart';

enum SessionStatus { idle, connecting, connected, failed }

class SessionState {
  const SessionState(this.status, [this.message]);
  final SessionStatus status;
  final String? message;

  static const idle = SessionState(SessionStatus.idle);
  static const connecting = SessionState(SessionStatus.connecting);
  static const connected = SessionState(SessionStatus.connected);
}

class RealtimeSession {
  RealtimeSession({required this.authProvider, required this.config});

  final AuthProvider authProvider;
  final SessionConfig config;

  final BackchannelPlayer _backchannelPlayer = BackchannelPlayer();
  WebRtcClient? _client;
  bool _interrupted = false;

  SessionState _state = SessionState.idle;
  SessionState get state => _state;

  void Function(SessionState state)? onStateChange;
  void Function(ServerEvent event)? onEvent;

  void _setState(SessionState next) {
    _state = next;
    onStateChange?.call(next);
  }

  Future<void> connect() async {
    if (_state.status != SessionStatus.idle &&
        _state.status != SessionStatus.failed) {
      return;
    }
    _setState(SessionState.connecting);
    try {
      final mic = await Permission.microphone.request();
      if (!mic.isGranted) {
        _setState(const SessionState(
          SessionStatus.failed,
          'Microphone access denied. Enable it in system Settings.',
        ));
        return;
      }

      final credentials = await authProvider.credentials();
      final api = SignalingApi(credentials);
      final iceServers = await api.fetchIceServers();

      final client = await WebRtcClient.create(iceServers);
      _client = client;
      _wireCallbacks(client);

      final offer = await client.makeOfferSdp();
      final answer = await api.postOffer(offer);
      await client.setAnswer(answer);
    } catch (error) {
      await disconnect(failure: error.toString());
    }
  }

  Future<void> disconnect({String? failure}) async {
    await _client?.close();
    _client = null;
    await _backchannelPlayer.stop();
    _interrupted = false;
    _setState(
      failure != null ? SessionState(SessionStatus.failed, failure) : SessionState.idle,
    );
  }

  void setMicEnabled(bool enabled) => _client?.setMicEnabled(enabled);

  void _wireCallbacks(WebRtcClient client) {
    client.onDataChannelOpen = _sendInitialEvents;
    client.onServerEvent = _handle;
    client.onConnectionStateChange = _handleConnectionState;
  }

  void _sendInitialEvents() {
    final client = _client;
    if (client == null) return;
    client.send(sessionUpdateEvent(config));
    client.send(conversationItemCreateEvent(config.greetingPrompt));
    client.send(responseCreateEvent());
    _setState(SessionState.connected);
  }

  void _handle(ServerEvent event) {
    switch (event) {
      case SpeechStarted():
        // Barge-in: silence the agent immediately, cancel its response.
        _interrupted = true;
        _client?.setAgentAudioEnabled(false);
        _client?.send(responseCancelEvent());
      case OutputItemAdded():
        if (_interrupted) {
          _client?.setAgentAudioEnabled(true);
          _interrupted = false;
        }
      case BackchannelAudioDelta(:final base64Pcm16):
        _backchannelPlayer.enqueue(base64Pcm16);
      case BackchannelAudioDone():
        _backchannelPlayer.flush();
      default:
        break;
    }
    onEvent?.call(event);
  }

  void _handleConnectionState(RTCPeerConnectionState pcState) {
    switch (pcState) {
      case RTCPeerConnectionState.RTCPeerConnectionStateFailed:
        disconnect(failure: 'Connection lost.');
      case RTCPeerConnectionState.RTCPeerConnectionStateDisconnected:
      case RTCPeerConnectionState.RTCPeerConnectionStateClosed:
        if (_state.status == SessionStatus.connected) disconnect();
      default:
        break;
    }
  }
}
